### PR TITLE
VP-6114: Load data from services if index object is null

### PR DESCRIPTION
**Pull request by Vectorfield4**
**Problem:**
We cannot work with a category or a product model if this model does not contain an indexed object.

**Solution:**
1. Load root object from services if the search module does not provide an indexed object inside it;
1. Add mapping `key` to `CatalogProduct` and `Category` in the catalog module: https://github.com/VirtoCommerce/vc-module-catalog/pull/448
1. Change mapping sequence by moving part of mapping which based on root object to finalization middleware;
1. Add some missing data to SearchResponces.

**Jira:**
https://virtocommerce.atlassian.net/browse/VP-6114

### CODE REVIEW

File: src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Middlewares/EnsureCatalogProductLoadingMiddleware.cs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        {
+            var itemsIds = parameter.Results
+                .Where(expProduct => expProduct.IndexedProduct is null)
+                .Select(expProduct => expProduct.Key)
+                .Where(key => key != null)
+                .ToArray();
+
+            if (itemsIds.Any())
+            {
+                var responceGroup = parameter.Query.GetResponseGroup();
```
Comment (yecli) [inline src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Middlewares/EnsureCatalogProductLoadingMiddleware.cs:29]: It should be another response group, you need to add the method which calculates `ItemResponseGroup` based on the query fields:
https://github.com/VirtoCommerce/vc-module-catalog/blob/bac3d033464b06eabecc3a9f8a5b1bcc799004f7/src/VirtoCommerce.CatalogModule.Core/Model/ItemResponseGroup.cs#L6

File: src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Middlewares/EnsureCategoryLoadingMiddleware.cs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        {
+            var itemsIds = parameter.Results
+                .Where(expProduct => expProduct.Category is null)
+                .Select(expProduct => expProduct.Key)
+                .Where(key => key != null)
+                .ToArray();
+
+            if (itemsIds.Any())
+            {
+                string responceGroup = null/*parameter.Query.GetResponseGroup()*/; // todo: responce group for category
```
Comment (yecli) [inline src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Middlewares/EnsureCategoryLoadingMiddleware.cs:29]: It should be another method and response group:
https://github.com/VirtoCommerce/vc-module-catalog/blob/bac3d033464b06eabecc3a9f8a5b1bcc799004f7/src/VirtoCommerce.CatalogModule.Core/Model/CategoryResponseGroup.cs#L6

**Review by yecli — CHANGES_REQUESTED** [review]

File: src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Middlewares/EnsureCatalogProductLoadingMiddleware.cs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        {
+            var itemsIds = parameter.Results
+                .Where(expProduct => expProduct.IndexedProduct is null)
+                .Select(expProduct => expProduct.Key)
+                .Where(key => key != null)
+                .ToArray();
+
+            if (itemsIds.Any())
+            {
+                var responceGroup = parameter.Query.GetItemResponseGroup();
```
Comment (yecli) [inline src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Middlewares/EnsureCatalogProductLoadingMiddleware.cs:29]: Typo
`responceGroup` =>`responseGroup`

**Review by yecli — COMMENTED** [review]
Checked with ElasticSearch and Store object in index = false. Probably it should work based on new middlewares - but no, got Null Reference exception. The reason probably is in incorrect mapping - `ExpProduct.Key` is empty. Object seems to be incorrect:
image

Example query:
```json
{
    product(
        id: "8b7b07c165924a879392f4f51a6f7ce0"
        storeId: "Electronics"
        userId: "d97ee2c7-e29d-440a-a43a-388eb5586087"
        cultureName: "en-us"
        currencyCode: "USD")
    {
        id
        name
        prices 
        {
         	list {
            amount
            currency {
              code
            }
          } 
         	sale {
            amount
            currency {
              code
            }
          } 
        }
    }
}
```

**Review by yecli — APPROVED** [review]

File: src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Mapping/ProductMappingProfile.cs [hunk 4]
```diff
@@ -28,59 +24,9 @@ public ProductMappingProfile()
...
-                }
-                var store = storeObj as StoreModule.Core.Model.Store;
-                context.Items.TryGetValue("language", out var language);
-                language ??= store.DefaultLanguage;
-
-                var genericModelBinder = new GenericModelBinder<ExpProduct>();
-                var result = genericModelBinder.BindModel(src) as ExpProduct;
-                if (result != null)
-                {
-                    result.AllPrices = context.Mapper.Map<IEnumerable<ProductPrice>>(result.IndexedPrices).ToList();
```
Comment (yecli) [inline src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Mapping/ProductMappingProfile.cs:45]: Is this code executed somewhere now? Or is it useless?

File: src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Queries/LoadRelatedCatalogOutlineQueryHandler.cs [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+            if (store is null) return null;
+
+            return new LoadRelatedCatalogOutlineResponse
+            {
+                Outline = request.Outlines.GetOutlinePath(store.Catalog)
+            };
+        }
+    }
+
+    public class LoadRelatedSlugPathQueryHandler : IQueryHandler<LoadRelatedSlugPathQuery, LoadRelatedSlugPathResponse>
```
Comment (yecli) [inline src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Queries/LoadRelatedCatalogOutlineQueryHandler.cs:30]: Let's move this class to standalone file

File: src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Mapping/ProductMappingProfile.cs (hunk 4 above)
Comment (Vectorfield4) [reply src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Mapping/ProductMappingProfile.cs:45]: > Is this code executed somewhere now?
In the second pipeline.
> Or is it useless?
Code duplication.

**Review by yecli — APPROVED** [review]

**Review by tatarincev — APPROVED** [review]

**Review by tatarincev — APPROVED** [review]

**Pull request merged**